src/Polysemy/Internal.hs

```diff
@@ -534,6 +540,143 @@ subsumeUsing pr =
+  r' ~ InsertedAfter e es r =>
+  Subsume r r' =>
+  Sem r a ->
+  Sem r' a
+insertAfter = subsume_
+{-# inline insertAfter #-}
+
+------------------------------------------------------------------------------
+-- | Finds the effect @e@ in the stack and inserts the effects @es@ before it.
+--
+-- @since TODO
+type family InsertedBefore (e :: Effect) (es :: EffectRow) (r :: EffectRow) :: EffectRow where
+  InsertedBefore e es (e : r) = Append es (e : r)
+  InsertedBefore e es (e0 : r) = e0 : InsertedBefore e es r
+
+------------------------------------------------------------------------------
+-- | Introduce a set of effects into 'Sem' before the position of the provided
+-- effect @e@. This is intended to be used with a type application:
+--
+-- @
+-- let
+--   sem1 :: Sem [e1, e2, e3, e4, e5] a
+--   sem1 = insertBefore @e5 @[e3, e4] (sem0 :: Sem [e1, e2, e5] a)
+-- @
+--
+-- @since TODO
+insertBefore ::
+  ∀ e es r r' a .
+  r' ~ InsertedBefore e es r =>
+  Subsume r r' =>
+  Sem r a ->
+  Sem r' a
+insertBefore = subsume_
+{-# inline insertBefore #-}
+
+
+
+data SList l where
+  SEnd  :: SList '[]
+  SCons :: SList xs -> SList (x ': xs)
+
+class KnownList l where
+  singList :: SList l
+
+instance KnownList '[] where
+  singList = SEnd
+  {-# INLINE singList #-}
+
+instance KnownList xs => KnownList (x ': xs) where
+  singList = SCons singList
+  {-# INLINE singList #-}
+
+class Tide (n :: Nat) l where
+  singTide :: SList l
+
+instance {-# OVERLAPPING #-} (l ~ '[]) => Tide 0 l where
+  singTide = SEnd
+
+instance (Tide (n - 1) xs, l ~ (x ': xs)) => Tide n l where
+  singTide = SCons (singTide @(n - 1))
+
+class Build (l :: [a]) (s :: [a]) (r :: [a]) (ix :: Nat) (head :: [a]) (returnType :: [a]) where
+  singBuild :: SList l
+
+instance (l ~ '[]) => Build l r r ix head return where
+  singBuild = SEnd
+
+instance {-# INCOHERENT #-} (Build l1 s r ix head return, l ~ (e ': l1))
+      => Build l (e ': s) r ix head return where
+  singBuild = SCons (singBuild @_ @l1 @s @r @ix @head @return)
+
+instance {-# INCOHERENT #-} TypeError (InsertAtFailure r ix head return)
+       => Build l s r ix head return where
+  singBuild = undefined
+
+type family InsertAtUnprovidedIndex where
+  InsertAtUnprovidedIndex = TypeError ('Text "insertAt: index must be provided")
+
+------------------------------------------------------------------------------
+-- | Introduce a set of effects into 'Sem' at the index @i@, before the effect
+-- that previously occupied that position. This is intended to be used with a
+-- type application:
+--
+-- @
+-- let
+--   sem1 :: Sem [e1, e2, e3, e4, e5] a
+--   sem1 = insertAt @2 @[e3, e4] (sem0 :: Sem [e1, e2, e5] a)
+-- @
+--
+-- @since TODO
+insertAt
+  :: forall n new top r s a
+   . ( Tide n top
+     , WhenStuck n InsertAtUnprovidedIndex
+     , s ~ Append new r, Build new s r n top (Append top s)
+     , Subsume (Append top r) (Append top s))
+  => Sem (Append top r) a
+  -> Sem (Append top s) a
+insertAt =
+  subsume_
```

Comment:
Remember that `subsume_` won't work here. You need to steal [`weakenMid`](https://hackage.haskell.org/package/in-other-words-0.2.0.0/docs/src/Control.Effect.Internal.KnownList.html#weakenMid) from IOW and use that, instead.

Implementation should be something like this (haven't doublechecked):
```
insertAt = hoistSem $ \u -> hoist (insertAt @n @new @top @r @s) $
  weakenMid @r (singTide @n @top) (singBuild @_ @new @s @r @n @top @(Append top s)) u
```
Some type applications may be redundant.